From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Vincent Strubel <clipos@ssi.gouv.fr>
Date: Tue, 14 Oct 2008 13:50:16 +0200
Subject: [PATCH 19/32] clip: vserver: Deny acces to the network without a SP

Add a 'NO_SP' network context flag to vserver. Packets can only be
emitted or received without a SP from the ADMIN context, and those
non-ADMIN contexts that have the NO_SP flag.

Patch: 1519_vserver_nx_nosp.patch
---
 include/uapi/vserver/network.h |  2 ++
 net/xfrm/xfrm_policy.c         | 15 ++++++++++++---
 2 files changed, 14 insertions(+), 3 deletions(-)

diff --git a/include/uapi/vserver/network.h b/include/uapi/vserver/network.h
index 581e42f8e770e..e9f53a4b392b5 100644
--- a/include/uapi/vserver/network.h
+++ b/include/uapi/vserver/network.h
@@ -24,6 +24,8 @@
 #define NXF_SC_HELPER		(1ULL << 36)
 #define NXF_PERSISTENT		(1ULL << 38)
 
+#define NXF_NO_SP		(1ULL << 48)
+
 #define NXF_ONE_TIME		(0x0005ULL << 32)
 
 
diff --git a/net/xfrm/xfrm_policy.c b/net/xfrm/xfrm_policy.c
index 65a41e657a7ea..377d1ca2c09ce 100644
--- a/net/xfrm/xfrm_policy.c
+++ b/net/xfrm/xfrm_policy.c
@@ -33,6 +33,8 @@
 #include <net/snmp.h>
 #endif
 
+#include <linux/vserver/debug.h>
+
 #include "xfrm_hash.h"
 
 #define XFRM_QUEUE_TMO_MIN ((unsigned)(HZ/10))
@@ -2244,14 +2246,21 @@ struct dst_entry *xfrm_lookup(struct net *net, struct dst_entry *dst_orig,
 		xflo.flags = flags;
 
 		/* To accelerate a bit...  */
-		if ((dst_orig->flags & DST_NOXFRM) ||
-		    !net->xfrm.policy_count[XFRM_POLICY_OUT])
+		if ((dst_orig->flags & DST_NOXFRM))
 			goto nopol;
 
 		flo = flow_cache_lookup(net, fl, family, dir,
 					xfrm_bundle_lookup, &xflo);
-		if (flo == NULL)
+		if (flo == NULL) {
+			if (sk && sk->sk_nx_info && !(sk->sk_nx_info->nx_flags & NXF_NO_SP)) {
+				vxwprintk_task(VX_WARNLEVEL,
+						"denied network access without a SP");
+				err = -EPERM;
+				goto dropdst;
+			}
 			goto nopol;
+		}
+
 		if (IS_ERR(flo)) {
 			err = PTR_ERR(flo);
 			goto dropdst;
-- 
2.17.0

